File: LeetCode/9.cpp
// LeetCode - 9. Palindrome Number

class Solution {
public:
    bool isPalindrome(int x) {
        if (x == 0)
            return true;
        if (x < 0 || x % 10 == 0)
            return false;
        int y = x;
        long rev = 0; // to avoid overflow
        while (y > 0) { // reverses the number
            rev *= 10;
            rev += y % 10;
            y /= 10;
        }
        return rev == x;
    }
};

File: LeetCode/9.ml
(* LeetCode - 9. Palindrome Number *)

let isPalindrome x = 
    if x = 0 
      then true
    else if (x < 0 || x mod 9 = 0)
      then false
    else
      let rev i = 
        let rec rev_aux n = function
          | 0 -> n
          | z -> rev_aux (n * 10 + (z mod 10)) (z / 10) in
        rev_aux 0 i in
      x = rev x